sktime/forecasting/base/_base.py

```diff
@@ -631,6 +637,48 @@ def update_predict_single(
             alpha=alpha,
         )
 
+    def predict_residuals(self, y, X=None, fh=None):
```

Comment:
Shouldn't `y` be optional for in-sample predictions?

Reply:
this is not necessarily in-sample, but we could make it that if no `y` is passed, in-sample residuals are computed, and the already seen `_y` are used. If that's convenient?

Reply:
So just confirming --- 

## For in sample residuals, the call would simply be 

```python
# Option 1: All In-sample residuals
forecaster.predict_residuals()  

# Last "n" (n=4 in this case) in-sample residuals
forecaster.predict_residuals(fh=[0, -1, -2, -3]) 
```

## For out of sample, it would be 

```python
# Option 1: # fh would be same length as Sth
forecaster.predict_residuals(y=Sth)  

# Option 2
# Would non-matching indices (between y and predictions) be ignored or raise error or warning?
forecaster.predict_residuals(y=Sth1, fh=Sth2)  
```

Reply:
ok, here's with what I'm currently thinking.

if you want in-sample residuals:
```python
# if fh must be passed in fit
forecaster.fit(y, fh=y.index).predict_residuals()
# if fh can be passed in predict
forecaster.fit(y).predict_residuals()
```

for out-of-sample residuals:
```python
# if fh must be passed in fit
forecaster.fit(y, fh=my_fh).predict_residuals(y_true)
# if fh can be passed in predict
forecaster.fit(y).predict_residuals(y_true)
```

The `fh` is inferred in either case, it's the inde of `y` in `fit` (in-sample) or `y_true` (out-of-sample); if no `y` is passed, the `y` in `fit` is used as the argument in `predict_residuals`.

Reply:
Hi @fkiraly Sorry for the delayed response. I was out over the holidays. I think your alternate solution is cleaner and we can work with this.

Reply:
alright!

Reply:
I'll write tests then so it's ready to go in.

Reply:
(done)